### PR TITLE
feat: added support for groth16

### PR DESCRIPTION
**Pull request by srdtrk**
closes: #119, #107

This also updates dependencies and includes a major refactor

### CODE REVIEW

File: contracts/src/SP1ICS07Tendermint.sol [hunk 1]
```diff
@@ -421,33 +433,41 @@ contract SP1ICS07Tendermint is
     }
 
     /// @notice Validates the Env public values.
-    /// @param env The public values.
-    function validateEnv(Env memory env) private view {
-        if (env.now > block.timestamp) {
-            revert ProofIsInTheFuture(block.timestamp, env.now);
+    /// @param publicClientState The public client state.
+    /// @param time The time.
+    function validateEnv(ClientState memory publicClientState, uint64 time) private view {
```
Comment (gjermundgaraba) [inline contracts/src/SP1ICS07Tendermint.sol:438]: A rename might be in order. validateEnv feels a bit weird here now.

File: contracts/src/SP1ICS07Tendermint.sol [hunk 2]
```diff
@@ -421,33 +433,41 @@ contract SP1ICS07Tendermint is
...
+                publicClientState.trustLevel.numerator,
+                publicClientState.trustLevel.denominator
             );
         }
-        if (env.trustingPeriod != clientState.trustingPeriod) {
-            revert TrustingPeriodMismatch(clientState.trustingPeriod, env.trustingPeriod);
+        if (publicClientState.trustingPeriod != clientState.trustingPeriod) {
+            revert TrustingPeriodMismatch(clientState.trustingPeriod, publicClientState.trustingPeriod);
         }
-        if (env.trustingPeriod > clientState.unbondingPeriod) {
```
Comment (gjermundgaraba) [inline contracts/src/SP1ICS07Tendermint.sol:449]: Why was this removed? Would we catch this anywhere else now?

File: contracts/test/Misbehaviour.t.sol [hunk 3]
```diff
@@ -42,14 +39,13 @@ contract SP1ICS07MisbehaviourTest is SP1ICS07TendermintTest {
 
         submitMsg = abi.decode(fixture.submitMsg, (IMisbehaviourMsgs.MsgSubmitMisbehaviour));
         output = abi.decode(submitMsg.sp1Proof.publicValues, (IMisbehaviourMsgs.MisbehaviourOutput));
-        env = output.env;
     }
 
     function test_ValidDoubleSignMisbehaviour() public {
-        setUpMisbehaviour("misbehaviour_double_sign_fixture.json");
+        setUpMisbehaviour("misbehaviour_double_sign-plonk_fixture.json");
```
Comment (gjermundgaraba) [inline contracts/test/Misbehaviour.t.sol:45]: super duper nit, but why is this file not named "misbehaviour_double_sign_fixture-plonk.json" to match the other files? :D

File: e2e/interchaintestv8/sp1_ics07_test.go [hunk 4]
```diff
@@ -53,7 +54,7 @@ type SP1ICS07TendermintTestSuite struct {
 
 // SetupSuite calls the underlying SP1ICS07TendermintTestSuite's SetupSuite method
 // and deploys the SP1ICS07Tendermint contract
-func (s *SP1ICS07TendermintTestSuite) SetupSuite(ctx context.Context) {
+func (s *SP1ICS07TendermintTestSuite) SetupSuite(ctx context.Context, opArgs ...string) {
```
Comment (gjermundgaraba) [inline e2e/interchaintestv8/sp1_ics07_test.go:57]: I'm not too fond of this approach. The test setup is not very easy to understand and we are leaking the operator parameters out to all the tests. I think it would be better if we just took in a prover enum or something and converted that to the correct arguments when needed.

**Comment by gjermundgaraba** [comment]
A general comment: It would have been nice if this had been split into two PRs ideally. It was a bit too large. 

(PS. I know I am a sinner in this aspect, but a general concept I want us all to get better at, so just putting it here :))

File: contracts/src/SP1ICS07Tendermint.sol (hunk 2 above)
Comment (srdtrk) [reply contracts/src/SP1ICS07Tendermint.sol:449]: I think this would be handled in the constructor now since you can't even create a client in this manner

File: contracts/test/Misbehaviour.t.sol (hunk 3 above)
Comment (srdtrk) [reply contracts/test/Misbehaviour.t.sol:45]: Because the way e2e's generate fixtures by appending `_fixture` to the end of the fixture name. I think this is fine because misbehavior fixtures are generated by a different process. So it reflects some implementation detail

File: contracts/src/SP1ICS07Tendermint.sol (hunk 2 above)
Comment (gjermundgaraba) [reply contracts/src/SP1ICS07Tendermint.sol:449]: That would make sense. Do we have a test that covers that scenario (if not, we should create an issue to do that)?

File: contracts/test/Misbehaviour.t.sol (hunk 3 above)
Comment (gjermundgaraba) [reply contracts/test/Misbehaviour.t.sol:45]: Fair enough :)

File: contracts/src/SP1ICS07Tendermint.sol (hunk 2 above)
Comment (srdtrk) [reply contracts/src/SP1ICS07Tendermint.sol:449]: We don't have any constructor tests, so perhaps an issue for that?

File: contracts/src/SP1ICS07Tendermint.sol (hunk 1 above)
Comment (srdtrk) [reply contracts/src/SP1ICS07Tendermint.sol]: Renamed. Feel free to change the name in the future if you don't like the new name.

File: contracts/src/SP1ICS07Tendermint.sol (hunk 2 above)
Comment (gjermundgaraba) [reply contracts/src/SP1ICS07Tendermint.sol:449]: Added an issue: https://github.com/cosmos/sp1-ics07-tendermint/issues/129

**Review by gjermundgaraba — APPROVED** [review]
